cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -413,12 +482,57 @@ pub mod tests {
             check_build_metadata(&env, EXAMPLE_PACKAGE_NAME, &flox.system).unwrap(),
         );
 
-        let publish_provider = PublishProvider::<&FloxBuildMk> {
+        let publish_provider = PublishProvider::<&FloxBuildMk, &MockCache> {
             build_metadata,
             env_metadata,
+            cache: None,
             _builder: None,
         };
 
-        let _res = publish_provider.publish(&client, &catalog_name);
+        let _res = publish_provider.publish(&client, &catalog_name).await;
+
+        assert!(_res.is_ok());
+    }
+
+    #[tokio::test]
+    async fn test_upload_to_cache() {
+        let (flox, _temp_dir_handle) = flox_instance();
+        let (_tempdir_handle, _remote_repo, remote_uri) = example_remote();
+        let (mut env, _build_repo) = example_path_environment(&flox, Some(&remote_uri));
+
+        // Do the build to ensure it's been run.  We just want to find the outputs
+        assert_build_status(&flox, &mut env, EXAMPLE_PACKAGE_NAME, true);
+
+        let client = Client::Mock(MockClient::new(None::<String>).unwrap());
+        let token = create_test_token("test");
+        let catalog_name = token.handle().to_string();
+
+        let (env_metadata, build_metadata) = (
+            check_environment_metadata(&flox, &mut env).unwrap(),
+            check_build_metadata(&env, EXAMPLE_PACKAGE_NAME, &flox.system).unwrap(),
+        );
+
+        // It's a lot easier to detect an error than a no-op success, so we test that case here.
+        let cache = Some(MockCache {
+            uri: Url::parse("s3://my-cool-cache").unwrap(),
+            error_msg: Some("Something went wrong".to_string()),
+        });
+
+        let publish_provider = PublishProvider::<&FloxBuildMk, &MockCache> {
+            build_metadata,
+            env_metadata,
+            cache: cache.as_ref(),
+            _builder: None,
+        };
+
+        let _res = publish_provider.publish(&client, &catalog_name).await;
+        assert!(!_res.is_ok());
+
+        if let Err(e) = _res {
+            assert_eq!(
+                e.to_string(),
+                "Failed to upload to cache: Something went wrong"
+            );
+        }
```

Comment:
although, doesn't that mainly test that the mock works?

given the name `test_upload_to_cache` i'd have expected this to publish to a `file://<tempdir>` store at least.

```suggestion
        let res = publish_provider.publish(&client, &catalog_name).await;
        let err = res.expect_err("publish should fail");
        assert_eq!(
            err.to_string(),
            "Failed to upload to cache: Something went wrong"
        );
```